### ISSUE TITLE
XMLHttpRequest (CORS) exception on every flutter web project

### ISSUE THREAD

**Issue by ahmadhassan7** [flutter/flutter#119297]
## Steps to Reproduce

Call any API in flutter web they all will give `XMLHttpRequest` error. Only way to deal with it is by flagging the security of chrome `flutter run -d chrome --web-browser-flag "--disable-web-security"` but this is not the solution and it doesn't works in release.
 
Adding CORS to headers also doesn't works. Same API has to be called by some middleware api then it works but upon directly calling the API to flutter web it always gives `XMLHttpRequest` error. 

Even after building flutter web and hosting it somewhere, `XMLHttpRequest` error doesn't goes away. 

**Expected results:** 

Api request should be successful.

**Actual results:** 

`Access to XMLHttpRequest at 'api url here' from origin 'http://localhost:49303' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

<details>
  <summary>Logs</summary>

```
[        ] 💪 Running with sound null safety 💪
[   +1 ms] 🔥  To hot restart changes while running, press "r" or "R".
[        ] For a more detailed help message, press "h". To quit, press "q".
[        ] An Observatory debugger and profiler on Chrome is available at: http://127.0.0.1:50027/CizzoLhRCnQ=
[+1573 ms] null
[ +112 ms] The Flutter DevTools debugger and profiler on Chrome is available at: http://127.0.0.1:9102?uri=http://127.0.0.1:50027/CizzoLhRCnQ=
[+1808 ms] XMLHttpRequest error.

**Comment by darshankawar** [comment]
@ahmadhassan7 
Looking at the error, check [this](https://github.com/flutter/flutter/issues/51125#issuecomment-589256586) comment for your reference and based on it, I'll close it.  If you disagree, write in comments and I'll reopen it.

**Comment by ahmadhassan7** [comment]
I've checked tons of stackoverflow questions and issues regarding this. 
Same 3rd party API's work directly in web frameworks. I've tried in vue.js and svelte.js they work fine so that shows its not some fundamental limitation of the web instead its just on flutter web.

**Comment by mohsinnaqvi606** [comment]
Yes, the issue is with flutter web. if it is server error than why we are getting same error by calling google places Api?

**Comment by darshankawar** [comment]
Reopening this for further investigation.

**Comment by darshankawar** [comment]
I am going ahead and labeling this issue for team's thoughts on expected behavior.

For team's reference: another issue with similar report: https://github.com/flutter/flutter/issues/120822

**Comment by yjbanov** [comment]
Is it an XHR that you are creating yourself? If so, can you please provide a sample that we could take a look at?

Or does this happen in every Flutter app, even in the counter app? All our tests that use `flutter run -d chrome` are currently passing, and it's working for me right now. Is there anything special about your environment?

**Comment by ahmadhassan7** [comment]
Yes XHR and it only comes upon consuming 3rd party API's in Flutter web app. All other project which doesn't uses any sort of API works flawless. 

My environment is clean, nothing added/edited. I'm using Flutter Version 3.3.10 arm64 Architecture.

One of the codes I was testing on is attached I'm calling [GoDaddy's](https://developer.godaddy.com/doc/endpoint/domains) API in it and similarly I'd ran tests on [Pexels](https://www.pexels.com/api/) and [Stripe](https://stripe.com/docs/development) API's too all resulting the same.
```
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

class DomainAvailability {
  final String domainName;
  final bool isAvailable;

  DomainAvailability({required this.domainName, required this.isAvailable});
}

class DomainAvailabilityService {
  static const String _apiUrl =
      'https://api.ote-godaddy.com/v1/domains/available';

  Future<DomainAvailability> checkAvailability(String domainName) async {
    final url = '$_apiUrl?domain=$domainName';
    final response = await http.get(Uri.parse(url), headers: {
      'Authorization':
          'sso-key 3mM44UcgfE8M7d_SGrAhFrzGoTDkPKR3jLeWa:Sj2bjsqAWKsji28w8JEdZd'
    });

    if (response.statusCode == 200) {
      final Map<String, dynamic> data = jsonDecode(response.body);
      return DomainAvailability(
          domainName: domainName, isAvailable: data['available']);
    } else {
      throw Exception('Failed to check domain availability');
    }
  }
}

class MainScreen extends StatefulWidget {
  const MainScreen({super.key});

  @override
  MainScreenState createState() => MainScreenState();
}

class MainScreenState extends State<MainScreen> {
  final _domainController = TextEditingController();
  bool _isLoading = false;
  DomainAvailability? _domainAvailability;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Check Domain Availability')),
      body: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            TextField(
              controller: _domainController,
              decoration: const InputDecoration(labelText: 'Enter domain name'),
            ),
            const SizedBox(height: 16),
            _isLoading
                ? const Center(child: CircularProgressIndicator())
                : ElevatedButton(
                    onPressed: _checkAvailability,
                    child: const Text('Check availability'),
                  ),
            if (_domainAvailability != null)
              Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const SizedBox(height: 16),
                  Text('Domain name: ${_domainAvailability!.domainName}'),
                  const SizedBox(height: 16),
                  Text(
                      'Is available: ${_domainAvailability!.isAvailable ? 'yes' : 'no'}'),
                ],
              ),
          ],
        ),
      ),
    );
  }

  void _checkAvailability() async {
    setState(() {
      _isLoading = true;
    });

    final domainName = _domainController.text;
    final domainAvailability =
        await DomainAvailabilityService().checkAvailability(domainName);

    setState(() {
      _isLoading = false;
      _domainAvailability = domainAvailability;
    });
  }

  @override
  void dispose() {
    _domainController.dispose();
    super.dispose();
  }
}
```

**Comment by ditman** [comment]
@ahmadhassan7 you claim the following:

> Same 3rd party API's work directly in web frameworks. I've tried them myself in vue.js and svelte.js they work fine so that shows its not some fundamental limitation of the web instead its just on flutter web.

Can you give me an example of a 3rd party API that worked for you in JS, but doesn't work in Flutter Web?

**Comment by ditman** [comment]
> Yes, the issue is with flutter web. if it is server error than why we are getting same error by calling google places Api?

@mohsinnaqvi606 can you provide any code of how you're using the Google Places API that reproduces this CORS issue?

**Comment by ditman** [comment]
PS: Remember you may need to use [`withCredentials`](https://pub.dev/documentation/http/latest/browser_client/BrowserClient/withCredentials.html) in the client from `package:http` for your `Authorization` header to be sent with your request.
